HeterogeneousCore/AlpakaMath/interface/Vector.h

```diff
@@ -0,0 +1,301 @@
+
+namespace cms::alpakatools::math {
+
+  template <typename T>
+  constexpr std::enable_if_t<std::is_arithmetic_v<T>, T> zero_() {
+    return static_cast<T>(0);
+  }
+
+  template <typename T, typename U>
+  constexpr std::enable_if_t<std::is_arithmetic_v<T> and std::is_arithmetic_v<U>, T> set_(U x) {
+    return static_cast<T>(x);
+  }
+
+  template <class T, int maxSize>
+  class Vector {
+  public:
+    // same notations as std::vector/array
+    using value_type = T;
+    static constexpr int N = maxSize;
+
+    constexpr int push_back_unsafe(const T &element) {
+      if (m_size < maxSize) {
+        const auto prev_size = m_size;
+        m_data[m_size] = element;
+        m_size += 1;
+        return prev_size;
+      }
+      return -1;
+    }
+
+    template <class... Ts>
+    constexpr int emplace_back_unsafe(Ts &&...args) {
+      if (m_size < maxSize) {
+        const auto prev_size = m_size;
+        m_data[m_size] = T(std::forward<Ts>(args)...);
+        m_size += 1;
+        return prev_size;
+      }
+      return -1;
+    }
+
+    constexpr T const &back() const {
+      assert(m_size > 0 && "Vector::back() on empty vector");
+      return m_data[m_size - 1];
+    }
+
+    constexpr T &back() {
+      assert(m_size > 0 && "Vector::back() on empty vector");
+      return m_data[m_size - 1];
+    }
+
+    // thread-safe version of the vector, when used in a kernel
+    template <typename TAcc>
+    ALPAKA_FN_ACC int push_back(const TAcc &acc, const T &element) {
+      const auto previousSize = alpaka::atomicAdd(acc, &m_size, 1, alpaka::hierarchy::Blocks{});
+      if (previousSize < maxSize) {
+        m_data[previousSize] = element;
+        return previousSize;
+      } else {
+        alpaka::atomicSub(acc, &m_size, 1, alpaka::hierarchy::Blocks{});
+        return -1;
+      }
+    }
+
+    template <typename TAcc, class... Ts>
+    ALPAKA_FN_ACC int emplace_back(const TAcc &acc, Ts &&...args) {
+      const auto previousSize = alpaka::atomicAdd(acc, &m_size, 1, alpaka::hierarchy::Blocks{});
+      if (previousSize < maxSize) {
+        m_data[previousSize] = T(std::forward<Ts>(args)...);
+        return previousSize;
+      } else {
+        alpaka::atomicSub(acc, &m_size, 1, alpaka::hierarchy::Blocks{});
+        return -1;
+      }
+    }
+
+    constexpr T pop_back() {
+      if (m_size > 0) {
+        auto previousSize = m_size--;
+        return m_data[previousSize - 1];
+      } else
+        return T();
+    }
+
+    constexpr Vector() : m_data{}, m_size(maxSize) {};
+
+    constexpr explicit Vector(const T &value) : m_data{}, m_size(maxSize) {
+      CMS_UNROLL_LOOP
+      for (int i = 0; i < maxSize; i++) {
+        m_data[i] = value;
+      }
+    }
+
+    template <typename... U,
+              typename = std::enable_if_t<(sizeof...(U) == maxSize) and (std::conjunction_v<std::is_same<T, U>...>)>>
+    constexpr Vector(U... args) : m_data{args...}, m_size(sizeof...(U)) {}
```

Comment:
I guess this could be expressed along
```suggestion
    template <typename... U>
    requires requires { sizeof...(U) == maxSize and std::conjunction_v<std::is_same<T, U>...> }
    constexpr Vector(U... args) : m_data{args...} {}
```
(leaving out the `m_size` because of my other comment)